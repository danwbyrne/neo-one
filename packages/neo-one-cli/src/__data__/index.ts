export * from './getClients';
export * from './getContracts';
export * from './buildTests';
