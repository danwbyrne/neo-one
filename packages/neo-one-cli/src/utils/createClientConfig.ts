import { Config } from '@neo-one/server-plugin';
import { Paths } from 'env-paths';

export interface ClientConfig {
  readonly paths: Omit<Paths, 'log'>;
}

export const createClientConfig = ({ paths }: ClientConfig): Config<ClientConfig> =>
  new Config({
    name: 'client',
    configPath: paths.config,
    defaultConfig: {
      paths,
    },
    schema: {
      type: 'object',
      required: ['paths'],
      properties: {
        paths: {
          type: 'object',
          required: ['data', 'config', 'cache', 'temp'],
          properties: {
            data: { type: 'string' },
            config: { type: 'string' },
            cache: { type: 'string' },
            temp: { type: 'string' },
          },
        },
      },
    },
  });
