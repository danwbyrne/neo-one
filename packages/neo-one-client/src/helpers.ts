import { common, crypto, scriptHashToAddress as scriptHashToAddressBase } from '@neo-one/client-core';
import { entropyToMnemonic, generateMnemonic, mnemonicToEntropy, validateMnemonic } from 'bip39';
import { AddressString, PrivateKeyString, PublicKeyString } from './types';

export const publicKeyToScriptHash = (publicKey: PublicKeyString): string =>
  common.uInt160ToString(crypto.publicKeyToScriptHash(common.stringToECPoint(publicKey)));

export const publicKeyToAddress = (publicKey: PublicKeyString): AddressString =>
  crypto.scriptHashToAddress({
    addressVersion: common.NEO_ADDRESS_VERSION,
    scriptHash: crypto.publicKeyToScriptHash(common.stringToECPoint(publicKey)),
  });

export const scriptHashToAddress = scriptHashToAddressBase;

export const addressToScriptHash = (address: AddressString): string =>
  common.uInt160ToString(
    crypto.addressToScriptHash({
      addressVersion: common.NEO_ADDRESS_VERSION,
      address,
    }),
  );

export const wifToPrivateKey = (wif: string): PrivateKeyString =>
  common.privateKeyToString(crypto.wifToPrivateKey(wif, common.NEO_PRIVATE_KEY_VERSION));

export const privateKeyToWIF = (privateKey: PrivateKeyString): string =>
  crypto.privateKeyToWIF(common.stringToPrivateKey(privateKey), common.NEO_PRIVATE_KEY_VERSION);

export const privateKeyToScriptHash = (privateKey: PrivateKeyString): string =>
  common.uInt160ToString(crypto.privateKeyToScriptHash(common.stringToPrivateKey(privateKey)));

export const privateKeyToAddress = (privateKey: PrivateKeyString): AddressString =>
  crypto.privateKeyToAddress({
    addressVersion: common.NEO_ADDRESS_VERSION,
    privateKey: common.stringToPrivateKey(privateKey),
  });

export const privateKeyToPublicKey = (privateKey: PrivateKeyString): PublicKeyString =>
  common.ecPointToString(crypto.privateKeyToPublicKey(common.stringToPrivateKey(privateKey)));

export const isNEP2 = (encryptedKey: string): boolean => crypto.isNEP2(encryptedKey);

export const encryptNEP2 = async ({
  password,
  privateKey,
}: {
  readonly password: string;
  readonly privateKey: PrivateKeyString;
}): Promise<string> =>
  crypto.encryptNEP2({
    addressVersion: common.NEO_ADDRESS_VERSION,
    privateKey: common.stringToPrivateKey(privateKey),
    password,
  });

export const decryptNEP2 = async ({
  password,
  encryptedKey,
}: {
  readonly password: string;
  readonly encryptedKey: string;
}): Promise<PrivateKeyString> => {
  const privateKey = await crypto.decryptNEP2({
    addressVersion: common.NEO_ADDRESS_VERSION,
    encryptedKey,
    password,
  });

  return common.privateKeyToString(privateKey);
};

export const createPrivateKey = (): PrivateKeyString => common.privateKeyToString(crypto.createPrivateKey());

export const createMnemonic = generateMnemonic;
export const mnemonicFromPrivateKey = entropyToMnemonic;
export const privateKeyFromMnemonic = mnemonicToEntropy;
export const isMnemonic = validateMnemonic;
