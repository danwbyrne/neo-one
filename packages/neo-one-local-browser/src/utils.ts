export const getSmartContractBasePath = (value: string) => `/node_modules/@neo-one/smart-contract/${value}`;
export const getSmartContractLibBasePath = (value: string) => `/node_modules/@neo-one/smart-contract-lib/${value}`;
export const getSmartContractPath = (value: string) => getSmartContractBasePath(`${value}`);
export const getSmartContractLibPath = (value: string) => getSmartContractLibBasePath(`${value}`);
export const CONTRACTS_PATH = '/one/contracts';
