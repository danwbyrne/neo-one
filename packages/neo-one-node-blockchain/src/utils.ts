import { common, InvalidFormatError, IOHelper, ScriptBuilder, TriggerType, VMState } from '@neo-one/client-common';
import {
  ApplicationEngine,
  BlockBase,
  CallFlags,
  ContractState,
  Notification,
  Transaction,
  utils as coreUtils,
  Verifiable,
  VM,
} from '@neo-one/node-core';
import { BN } from 'bn.js';
import { ScriptVerifyError } from './errors';

const hashListBatchSize = 2000;

const getOnPersistNativeContractScript = coreUtils.lazy(() => {
  const hashes = [common.nativeHashes.GAS, common.nativeHashes.NEO];
  const script = new ScriptBuilder();
  hashes.forEach((hash) => {
    script.emitAppCall(hash, 'onPersist');
    script.emitOp('DROP');
  });

  return script.build();
});

// tslint:disable-next-line: no-any
const isTransaction = (value: any): value is Transaction => value?.type === 'Transaction';

const getApplicationExecuted = (engine: ApplicationEngine, container?: Verifiable) => ({
  transaction: isTransaction(container) ? container : undefined,
  trigger: engine.trigger,
  state: engine.state,
  gasConsumed: engine.gasConsumed,
  stack: engine.resultStack,
  notifications: engine.notifications.map((item) => Notification.fromStackItem(item, container)),
});

const getCallReceipt = (engine: ApplicationEngine, container?: Verifiable) => ({
  state: engine.state,
  gasConsumed: engine.gasConsumed,
  stack: engine.resultStack,
  notifications: engine.notifications.map((item) => Notification.fromStackItem(item, container)),
});

const verifyContract = async (contract: ContractState, vm: VM, transaction: Transaction) => {
  const verify = contract.manifest.abi.getMethod('verify');
  if (verify === undefined) {
    throw new InvalidFormatError(`the smart contract ${contract.scriptHash} does not have a verify method`);
  }

  const init = contract.manifest.abi.getMethod('_initialize');
  const gas = vm.withApplicationEngine(
    {
      trigger: TriggerType.Verification,
      container: transaction,
      snapshot: 'clone',
      gas: new BN(0),
      testMode: true,
    },
    (engine) => {
      engine.loadScript(contract.script, CallFlags.None);
      engine.setInstructionPointer(verify.offset);
      if (init !== undefined) {
        engine.setInstructionPointer(init.offset);
      }

      engine.loadScript(Buffer.from([]), CallFlags.None);
      const result = engine.execute();

      if (result === VMState.FAULT) {
        throw new ScriptVerifyError(`contract ${contract.scriptHash} returned FAULT state`);
      }

      if (engine.resultStack.length !== 1 || !engine.resultStack[0].getBoolean()) {
        throw new ScriptVerifyError(`contract ${contract.scriptHash} returns false`);
      }

      return engine.gasConsumed;
    },
  );

  return { fee: gas, size: IOHelper.sizeOfUInt8 * 2 };
};

const blockComparator = <TBlock extends BlockBase>({ index: aIndex }: TBlock, { index: bIndex }: TBlock) => {
  if (aIndex > bIndex) {
    return 1;
  }

  if (aIndex < bIndex) {
    return -1;
  }

  return 0;
};

export const utils = {
  ...coreUtils,
  hashListBatchSize,
  getApplicationExecuted,
  getCallReceipt,
  getOnPersistNativeContractScript,
  verifyContract,
  blockComparator,
  isTransaction,
};
