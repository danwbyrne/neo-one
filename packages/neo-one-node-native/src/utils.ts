import { InvalidFormatError } from '@neo-one/client-common';
import {
  ArrayStackItem,
  assertPrimitiveStackItem,
  assertStackItemType,
  BinaryReader,
  BooleanStackItem,
  BufferStackItem,
  ByteStringStackItem,
  IntegerStackItem,
  MapStackItem,
  NullStackItem,
  PrimitiveStackItem,
  StackItem,
  StackItemType,
  StorageItem,
  StructStackItem,
} from '@neo-one/node-core';
import { BN } from 'bn.js';
import _ from 'lodash';

interface ContainerPlaceholderAdd {
  readonly type: StackItemType;
  readonly elementCount: number;
}

class ContainerPlaceholder {
  public readonly type: StackItemType;
  public readonly elementCount: number;
  public readonly isContainerPlaceHolder = true;

  public constructor({ type, elementCount }: ContainerPlaceholderAdd) {
    this.type = type;
    this.elementCount = elementCount;
  }
}

// tslint:disable-next-line: no-any
const isContainerPlaceholder = (item: any): item is ContainerPlaceholder => item.isContainerPlaceHolder === true;

const getInteroperable = <T>(item: StorageItem, fromStackItem: (item: StackItem) => T): T => {
  const buffer = item.value;
  const reader = new BinaryReader(buffer);
  const deserializedStackItem = deserialize(reader, 16, 34);

  return fromStackItem(deserializedStackItem);
};

// tslint:disable: no-array-mutation no-loop-statement increment-decrement
const deserialize = (reader: BinaryReader, maxArraySize: number, maxItemSize: number): StackItem => {
  const deserialized: Array<StackItem | ContainerPlaceholder> = [];
  let undeserialized = 1;
  while (undeserialized-- > 0) {
    const type = assertStackItemType(reader.readInt8());
    switch (type) {
      case StackItemType.Any:
        deserialized.unshift(new NullStackItem());
        break;

      case StackItemType.Boolean:
        deserialized.unshift(new BooleanStackItem(reader.readBoolean()));
        break;

      case StackItemType.Integer:
        deserialized.unshift(new IntegerStackItem(new BN(reader.readVarBytesLE(IntegerStackItem.maxSize))));
        break;

      case StackItemType.ByteString:
        deserialized.unshift(new ByteStringStackItem(reader.readVarBytesLE(maxItemSize)));
        break;

      case StackItemType.Buffer:
        const size = reader.readVarUIntLE(maxItemSize).toNumber();
        deserialized.unshift(new BufferStackItem(reader.readVarBytesLE(size)));
        break;

      case StackItemType.Array:
        const arrayCount = reader.readVarUIntLE(maxArraySize).toNumber();
        deserialized.push(new ContainerPlaceholder({ type, elementCount: arrayCount }));
        undeserialized += arrayCount;
        break;

      case StackItemType.Struct:
        const structCount = reader.readVarUIntLE(maxArraySize).toNumber();
        deserialized.push(new ContainerPlaceholder({ type, elementCount: structCount }));
        undeserialized += structCount;
        break;

      case StackItemType.Map:
        const mapCount = reader.readVarUIntLE(maxArraySize).toNumber();
        deserialized.push(new ContainerPlaceholder({ type, elementCount: mapCount }));
        undeserialized += mapCount * 2;
        break;

      default:
        throw new InvalidFormatError(`Invalid StackItemType, found: ${type}`);
    }
  }

  let stackTemp: StackItem[] = [];
  while (deserialized.length > 0) {
    const item = deserialized.pop();
    if (item === undefined) {
      throw new Error('Unexpected undefined since deserialized.length > 0');
    }
    // tslint:disable-next-line: no-any
    if (isContainerPlaceholder(item)) {
      switch (item.type) {
        case StackItemType.Array:
          const arrayElements = stackTemp.slice(0, item.elementCount);
          stackTemp = stackTemp.slice(item.elementCount);
          stackTemp.push(new ArrayStackItem(arrayElements));
          break;

        case StackItemType.Struct:
          const structElements = stackTemp.slice(0, item.elementCount);
          stackTemp = stackTemp.slice(item.elementCount);
          stackTemp.push(new StructStackItem(structElements));
          break;

        case StackItemType.Map:
          const mapElements = stackTemp.slice(0, item.elementCount * 2);
          stackTemp = stackTemp.slice(item.elementCount * 2);
          stackTemp.push(
            new MapStackItem(
              _.range(0, item.elementCount * 2, 2).reduce((acc, idx) => {
                const key = assertPrimitiveStackItem(mapElements[idx]);
                acc.set(key, mapElements[idx + 1]);

                return acc;
              }, new Map<PrimitiveStackItem, StackItem>()),
            ),
          );
          break;

        default:
          throw new InvalidFormatError(`Invalid ContainerPlaceholder type, found: ${item.type}`);
      }
    } else {
      stackTemp.push(item);
    }
  }

  return stackTemp[0];
};
// tslint:enable: no-array-mutation no-loop-statement increment-decrement

const getSerializableArray = <T>(
  value: Buffer,
  readValue: (reader: BinaryReader) => T,
  max: number | BN = 0x1000000,
): readonly T[] => {
  const reader = new BinaryReader(value);

  return reader.readArray(() => readValue(reader), max);
};

const getSerializableArrayFromStorageItem = <T>(
  item: StorageItem,
  readValue: (reader: BinaryReader) => T,
  max: number | BN = 0x1000000,
): readonly T[] => {
  const value = item.value;

  return getSerializableArray(value, readValue, max);
};

export const utils = {
  getInteroperable,
  getSerializableArray,
  getSerializableArrayFromStorageItem,
};
