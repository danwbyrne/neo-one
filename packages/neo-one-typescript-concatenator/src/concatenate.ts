import { tsUtils } from '@neo-one/ts-utils';
import * as fs from 'fs-extra';
import path from 'path';
import ts from 'typescript';
import { Concatenator } from './Concatenator';

export function concatenate(entry: string) {
  const servicesHost: ts.LanguageServiceHost = {
    getScriptFileNames: () => [entry],
    getScriptVersion: () => '0',
    getScriptSnapshot: (fileName) => {
      // tslint:disable-next-line no-non-null-assertion
      if (!servicesHost.fileExists!(fileName)) {
        return undefined;
      }

      // tslint:disable-next-line no-non-null-assertion
      return ts.ScriptSnapshot.fromString(servicesHost.readFile!(fileName)!);
    },
    getCurrentDirectory: () => process.cwd(),
    getCompilationSettings: () => {
      const configPath = ts.findConfigFile(entry, ts.sys.fileExists);
      if (configPath === undefined) {
        return ts.getDefaultCompilerOptions();
      }
      const text = fs.readFileSync(configPath, 'utf8');
      const parseResult = ts.parseConfigFileTextToJson(configPath, text);

      return ts.parseJsonConfigFileContent(
        parseResult.config,
        {
          useCaseSensitiveFileNames: true,
          readDirectory: ts.sys.readDirectory,
          fileExists: ts.sys.fileExists,
          readFile: ts.sys.readFile,
        },
        path.dirname(configPath),
        undefined,
        undefined,
      ).options;
    },
    getDefaultLibFileName: (opts) => ts.getDefaultLibFilePath(opts),
    useCaseSensitiveFileNames: () => ts.sys.useCaseSensitiveFileNames,
    getNewLine: () => ts.sys.newLine,
    fileExists: ts.sys.fileExists,
    readFile: ts.sys.readFile,
    readDirectory: ts.sys.readDirectory,
  };

  const languageService = ts.createLanguageService(servicesHost);
  const program = languageService.getProgram();
  if (program === undefined) {
    throw new Error('hmmm');
  }
  const typeChecker = program.getTypeChecker();

  const getSymbol = (node: ts.Node) => {
    const symbol = tsUtils.node.getSymbol(typeChecker, node);
    if (symbol === undefined) {
      return undefined;
    }

    const aliased = tsUtils.symbol.getAliasedSymbol(typeChecker, symbol);
    if (aliased !== undefined) {
      return aliased;
    }

    return symbol;
  };

  const isIgnoreFile = () => false;
  const isGlobalIdentifier = () => false;
  const isGlobalFile = () => false;
  const isGlobalSymbol = () => false;

  const sourceFile = tsUtils.file.getSourceFileOrThrow(program, entry);

  const concatenator = new Concatenator({
    context: {
      typeChecker,
      program,
      languageService,
      getSymbol,
      isIgnoreFile,
      isGlobalIdentifier,
      isGlobalFile,
      isGlobalSymbol,
    },
    sourceFile,
  });

  const sourceFiles = concatenator.sourceFiles;
  if (sourceFiles.length === 0 || sourceFiles.length === 1) {
    return undefined;
  }

  return tsUtils.printBundle(program, sourceFiles, concatenator.substituteNode);
}
