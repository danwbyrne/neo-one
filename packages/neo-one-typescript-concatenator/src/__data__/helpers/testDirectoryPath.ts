import { normalizePath } from '@neo-one/utils';
import appRootDir from 'app-root-dir';
import * as path from 'path';

export const testDirectoryPath = (directory: string) =>
  normalizePath(
    path.resolve(
      appRootDir.get(),
      'packages',
      'neo-one-typescript-concatenator',
      'src',
      '__data__',
      'test-directories',
      directory,
      'index.ts',
    ),
  );
