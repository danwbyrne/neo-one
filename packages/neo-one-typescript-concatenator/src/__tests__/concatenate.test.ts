import { normalizePath } from '@neo-one/utils';
import appRootDir from 'app-root-dir';
import * as path from 'path';
import { testDirectoryPath } from '../__data__/helpers';
import { concatenate } from '../concatenate';

const unitTests: ReadonlyArray<string> = [
  // 'baseline',
  // 'collisions',
  // 'baselineWithSubModules',
  // 'collisionsWithSubModules',
  'constImportExports',
];

describe.only('Concatenator Testing Suite Mk 1.3b.73', () => {
  const runConcatenateTest = (entry: string) => {
    test(`${entry} concatenation test`, () => {
      const entryPath = testDirectoryPath(entry);

      const result = concatenate(entryPath);
      if (result === undefined) {
        return;
      }
      expect(result).toMatchSnapshot();
    });
  };

  unitTests.forEach(runConcatenateTest);

  test.skip('THE BIG ONE (@neo-one/utils)', () => {
    const entryPath = normalizePath(path.resolve(appRootDir.get(), 'packages', 'neo-one-utils', 'src', 'index.ts'));

    const result = concatenate(entryPath);
    if (result === undefined) {
      return;
    }
    expect(result.text).toMatchSnapshot();
  });
});
