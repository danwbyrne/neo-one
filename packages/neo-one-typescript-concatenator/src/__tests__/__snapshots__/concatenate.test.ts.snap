// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Concatenator Testing Suite Mk 1.3b.73 constImportExports concatenation test 1`] = `
Object {
  "sourceMap": Object {
    "file": "index.ts",
    "mappings": "AAAO,MAAM,GAAG,GAAG,KAAK,CAAC;ACAzB,MAAM,GAAG,GAAG,KAAK,CAAC;AACX,MAAM,IAAI,GAAG,GAAG,CAAC;ACExB,OAAO,EAAE,GAAG,IAAI,GAAG,EAAE,CAAC;AACtB,OAAO,EAAE,IAAI,IAAI,GAAG,EAAE,CAAC",
    "names": Array [],
    "sourceRoot": "/Users/danielbyrne/neo-one/packages/neo-one-typescript-concatenator/src/__data__/test-directories/constImportExports/index.ts/",
    "sources": Array [
      "foo.ts",
      "fizz.ts",
      "index.ts",
    ],
    "sourcesContent": Array [
      "export const foo = 'foo';
",
      "const foo = 'bar';
export const fizz = foo;
",
      "import * as fizz from './fizz';
import * as foo from './foo';

export { foo as bar };
export { fizz as foo };
",
    ],
    "version": 3,
  },
  "text": "const sc1 = 'foo';
const sc2 = 'bar';
const fizz = sc2;
export { foo as bar };
export { fizz as foo };
",
}
`;
